answerking_app/tests/test_unit/test_unit_productserializer.py

```diff
@@ -0,0 +1,211 @@
+            name=self.test_prod_data["name"]
+        )
+        category: Category = test_prod.category_set.first()
+        test_serializer_data: ReturnDict = ProductSerializer(test_prod).data
+        actual_category: dict = test_serializer_data["categories"][0]
+
+        self.assertEqual(actual_category["id"], category.id)
+        self.assertEqual(actual_category["name"], category.name)
+        self.assertEqual(actual_category["description"], category.description)
+
+    def test_product_serializer_retired_field_content(self):
+        test_prod: Product = Product.objects.get(
+            name=self.test_prod_data["name"]
+        )
+        test_serializer_data: ReturnDict = ProductSerializer(test_prod).data
+        expected: bool = test_prod.retired
+        actual: bool = test_serializer_data["retired"]
+        self.assertEqual(actual, expected)
+
+    def test_product_serializer_name_max_length_fail(self):
+        serializer_data: dict = copy.deepcopy(self.test_prod_data)
+        serializer_data["name"] = "e" * 51
+        serializer: ProductSerializer = ProductSerializer(data=serializer_data)
+
+        self.assertFalse(serializer.is_valid())
+        self.assertEqual(set(serializer.errors), {"name"})
+
+    def test_product_serializer_name_blank_fail(self):
+        serializer_data: dict = copy.deepcopy(self.test_prod_data)
+        serializer_data["name"] = ""
+        serializer: ProductSerializer = ProductSerializer(data=serializer_data)
+
+        self.assertFalse(serializer.is_valid())
+        self.assertEqual(set(serializer.errors), {"name"})
+
+    def test_product_serializer_name_regex_validator_pass_1(self):
+        allowed_characters: str = "abcdefghijklm nopqrstuvwxyz"
+        serializer_data: dict = copy.deepcopy(self.test_prod_data)
+        serializer_data["name"] = allowed_characters
+        serializer = ProductSerializer(data=serializer_data)
+
+        self.assertTrue(serializer.is_valid())
+        self.assertEqual(serializer.data["name"], allowed_characters)
+
+    def test_product_serializer_name_regex_validator_pass_2(self):
+        allowed_characters: str = "ABCDEFGHIJKLMNOPQRSTUVWXYZ!"
+        serializer_data: dict = copy.deepcopy(self.test_prod_data)
+        serializer_data["name"] = allowed_characters
+        serializer = ProductSerializer(data=serializer_data)
+
+        self.assertTrue(serializer.is_valid())
+        self.assertEqual(serializer.data["name"], allowed_characters)
+
+    def test_product_serializer_name_regex_validator_fail(self):
+        allowed_characters: str = "abcdefghijklmnopqrstuvwxyz "
+        serializer_data: dict = copy.deepcopy(self.test_prod_data)
+        serializer_data["name"] = allowed_characters + "#"
+        serializer: ProductSerializer = ProductSerializer(data=serializer_data)
+
+        self.assertFalse(serializer.is_valid())
+        self.assertEqual(set(serializer.errors), {"name"})
+
+    def test_product_serializer_price_max_length_fail(self):
```

Comment:
this is actually breaking 2 validators; max digits and max value. I would probably change this to test the max validator only as this will always be triggered if max digits is greater than 18.

Reply:
yeah makes sense

Reply:
could we use MAXNUMBERSIZE +1 as a int rather than '1'*19 (MAXNUMBERSIZE = 2147483647). This way we know that it fails at any number greater than MAXNUMBERSIZE